mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -59,6 +59,166 @@ struct RockLinalgAlignPass
+      suffixProduct[groupSize - 1] = 1;
+      for (unsigned i = groupSize - 1; i > 0; i--)
+        suffixProduct[i - 1] =
+            suffixProduct[i] * collapseOp.getSrcType().getDimSize(groups[i]);
+      // Derive the index values along all dimensions of the source
+      // corresponding to the index wrt to collapsed shape op output.
+      SmallVector<AffineExpr, 4> srcIndexExpr = getDelinearizedAffineExpr(
+          suffixProduct, collapseOp.getSrcType().cast<ShapedType>().getShape(),
+          rewriter, iDimCount++);
+      for (unsigned i = 0; i < groupSize; i++) {
+        resultExprs.push_back(srcIndexExpr[i]);
+      }
+    }
+    auto representativeMap = AffineMap::get(
+        /*numDims=*/collapseOp.getType().cast<ShapedType>().getShape().size(),
+        /*numSymbols=*/0, resultExprs, rewriter.getContext());
+    foldedMap = representativeMap.compose(foldedMap);
+    toBeErasedViewLikeOps.push_back(collapseOp);
+    return foldViewLikeOperands(rewriter, collapseOp.getViewSource(), foldedMap,
+                                rootOp, toBeErasedViewLikeOps);
+  }
+  if (memref::ExpandShapeOp expandShapeOp =
+          op.getDefiningOp<memref::ExpandShapeOp>()) {
+    LLVM_DEBUG(llvm::dbgs() << "InlineViewLikeOperandsLinalgRewritePattern "
+                               "does not currently support expand shapes as it "
+                               "is unlikely encounter with drop unit dims\n");
+    LLVM_DEBUG(llvm::dbgs()
+               << "If needed, it would be similiar to add logic as above.\n");
+    return failure();
+  }
+  rootOp = op;
+  return success();
+}
+
+LogicalResult InlineViewLikeOperandsLinalgRewritePattern::matchAndRewrite(
+    linalg::GenericOp laGeneric, PatternRewriter &rewriter) const {
+  auto loc = laGeneric.getLoc();
+  int ioCount = 0;
+
+  SmallVector<AffineMap, 4U> newMaps;
+  SmallVector<Value, 4U> newInputs;
+  unsigned int changedIOCount = 0;
+
+  SmallVector<Operation *> toBeErasedViewLikeOps;
+  for (auto input : laGeneric.inputs()) {
+    AffineMap resMap = laGeneric.getIndexingMapsArray()[ioCount++];
+    Value rootOp;
+    if (foldViewLikeOperands(rewriter, input, resMap, rootOp,
+                             toBeErasedViewLikeOps)
+            .failed()) {
+      return failure();
+    }
+    if (rootOp != input)
+      changedIOCount++;
+    newInputs.push_back(rootOp);
+    newMaps.push_back(resMap);
+  }
+  SmallVector<Value, 4U> newOutputs;
+  for (auto output : laGeneric.outputs()) {
+    AffineMap resMap = laGeneric.getIndexingMapsArray()[ioCount++];
+    Value rootOp;
+    if (foldViewLikeOperands(rewriter, output, resMap, rootOp,
+                             toBeErasedViewLikeOps)
+            .failed()) {
+      return failure();
+    }
+    if (rootOp != output)
+      changedIOCount++;
+    newOutputs.push_back(rootOp);
+    newMaps.push_back(resMap);
+  }
+  if (changedIOCount == 0)
+    return failure();
+
+  SmallVector<StringRef> iteratorTypes = llvm::to_vector<4>(
```

Comment:
I take it `getInputMutable()` and friends didn't come through here? Or, in the case of the affine maps, `setAttr`?